dspace-api/src/main/java/org/dspace/identifier/VersionedDOIIdentifierProvider.java

```diff
@@ -0,0 +1,350 @@
+            throws IdentifierException
+    {
+        if (!(dso instanceof Item))
+        {
+            throw new IdentifierException("Currently only Items are supported for DOIs.");
+        }
+        Item item = (Item) dso;
+        
+        if (StringUtils.isEmpty(identifier))
+        {
+            identifier = mint(context, dso);
+        }
+        String doiIdentifier = doiService.formatIdentifier(identifier);
+        
+        DOI doi = null;
+
+        // search DOI in our db
+        try
+        {
+            doi = loadOrCreateDOI(context, dso, doiIdentifier);
+        } catch (SQLException ex) {
+            log.error("Error in databse connection: " + ex.getMessage());
+            throw new RuntimeException("Error in database conncetion.", ex);
+        }
+
+        if (DELETED.equals(doi.getStatus()) ||
+                TO_BE_DELETED.equals(doi.getStatus()))
+        {
+            throw new DOIIdentifierException("You tried to register a DOI that "
+                    + "is marked as DELETED.", DOIIdentifierException.DOI_IS_DELETED);
+        }
+
+        // Check status of DOI
+        if (IS_REGISTERED.equals(doi.getStatus()))
+        {
+            return;
+        }
+        
+        String metadataDOI = getDOIOutOfObject(dso);
+        if (!StringUtils.isEmpty(metadataDOI)
+                && !metadataDOI.equalsIgnoreCase(doiIdentifier))
+        {
+            // remove doi of older version from the metadata
+                try {
+                    removePreviousVersionDOIsOutOfObject(context, item, metadataDOI);
+                } catch (AuthorizeException ex) {
+                    throw new RuntimeException("Trying to remove an old DOI from a versioned item, but wasn't authorized to.", ex);
+                }
+        }
+        
+        // change status of DOI
+        doi.setStatus(TO_BE_REGISTERED);
+        try {
+            doiService.update(context, doi);
+        }
+        catch (SQLException ex)
+        {
+            log.warn("SQLException while changing status of DOI {} to be registered.", ex);
+            throw new RuntimeException(ex);
+        }
+    }
+    
+    protected String getBareDOI(String identifier)
+            throws DOIIdentifierException
+    {
+        doiService.formatIdentifier(identifier);
+        String doiPrefix = DOI.SCHEME.concat(getPrefix())
+                                     .concat(String.valueOf(SLASH))
+                                     .concat(getNamespaceSeparator());
+        String doiPostfix = identifier.substring(doiPrefix.length());
+        if (doiPostfix.matches(pattern) && doiPostfix.lastIndexOf(DOT) != -1)
+        {
+            return doiPrefix.concat(doiPostfix.substring(0, doiPostfix.lastIndexOf(DOT)));
+        }
+        // if the pattern does not match, we are already working on a bare handle.
+        return identifier;
+    }
+    
+    private String getDOIPostfix(String identifier) 
```

Comment:
Make this method protected, so that if custom code is needed the class inheriting from this one can make use it.